### PR TITLE
Integrate WooCommerce with Jetpack Sync

### PR DESCRIPTION
**Pull request by gravityrail**
Syncs selected Woo data (order items and order item meta) to WPCOM using Jetpack Sync.

To test:
- ensure there is a checkout of the [WooCommerce](https://github.com/woocommerce/woocommerce) plugin in the same directory as the Jetpack plugin. This needs to be the original source, not the published plugin, so that it includes test files.
- Run `JETPACK_TEST_WOOCOMMERCE=1 phpunit --testsuite=sync --filter=WP_Test_Jetpack_Sync_WooCommerce`

### CODE REVIEW

File: sync/class.jetpack-sync-actions.php [hunk 1]
```diff
@@ -340,3 +355,5 @@ static function init_sync_cron_jobs() {
 
 // We need to define this here so that it's hooked before `updating_jetpack_version` is called
 add_action( 'updating_jetpack_version', array( 'Jetpack_Sync_Actions', 'do_initial_sync' ), 10, 2 );
+
+add_action( 'plugins_loaded', array( 'Jetpack_Sync_Actions', 'setup_woocommerce' ) );
```
Comment (lezama) [inline sync/class.jetpack-sync-actions.php]: is `plugins_loaded` isn't "early" enough?

performance wise maybe it is better just to initialise the module/listeners even if woo isn't installed?

Comment (gravityrail) [inline sync/class.jetpack-sync-actions.php]: plugins_loaded is before init, and is the earliest we can detect whether woo has been enabled. Performance-wise, I don't think this adds much overhead. Which part do you think is expensive?

**Comment by allendav** [comment]
I think this expansion of data collection should be approved by WooCommerce merchants.  Let's at least be clear about this in the terms of service.  Also, how do we surface this change in behavior to existing Jetpack and WooCommerce users?

**Comment by ebinnion** [comment]
Pushing this to 4.5 since Pit Crew team is trying to get a beta out today.

**Comment by samhotchkiss** [comment]
@gravityrail -- I'm taking this out of the current queue until Allen's concerns are reviewed/addressed

**Comment by gravityrail** [comment]
Sure thing @samhotchkiss. Makes sense.

**Comment by gravityrail** [comment]
@beaulebens has given go-ahead to merge this @samhotchkiss - adding back to 4.5

**Comment by jeffstieler** [comment]
What is the plan to surface the expansion of data collection for WooCommerce merchants?

@samhotchkiss @beaulebens

**Comment by beaulebens** [comment]
The additional details will be included in the [Jetpack support doc](https://jetpack.com/support/what-data-does-jetpack-sync/), which is what lets Jetpack users know what data is being synced.

File: sync/class.jetpack-sync-module-woocommerce.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		add_action( 'woocommerce_payment_complete', $callable, 10, 1 );
+
+		// order items
+		add_action( 'woocommerce_new_order_item', $callable, 10, 4 );
+		add_action( 'woocommerce_update_order_item', $callable, 10, 4 );
+		add_filter( 'jetpack_sync_before_enqueue_woocommerce_new_order_item', array( $this, 'filter_order_item' ) );
+		add_filter( 'jetpack_sync_before_enqueue_woocommerce_update_order_item', array( $this, 'filter_order_item' ) );
+
+		// order item meta
+		$this->init_listeners_for_meta_type( 'order_item', $callable );
```
Comment (lezama) [inline sync/class.jetpack-sync-module-woocommerce.php]: should this use `private $meta_types`?

Comment (gravityrail) [inline sync/class.jetpack-sync-module-woocommerce.php]: That variable is unused so I removed it.

**Comment by dereksmart** [comment]
merged to 4.5 44c43c3

**Pull request merged**